Libraries/LibGfx/JPGLoader.cpp

```diff
@@ -540,13 +571,66 @@ static bool read_start_of_scan(BufferStream& stream, JPGLoadingContext& context)
-    // The three values should be fixed for baseline JPEGs utilizing sequential DCT.
-    if (spectral_selection_start != 0 || spectral_selection_end != 63 || successive_approximation != 0) {
-        dbg() << stream.offset() << ": ERROR! Start of Selection: " << spectral_selection_start
-              << ", End of Selection: " << spectral_selection_end
-              << ", Successive Approximation: " << successive_approximation << "!";
+
+    u8 successive_approximation_high_bits = successive_approximation >> 4;
+    u8 successive_approximation_low_bits = successive_approximation & 0x0F;
+
+    u8 selection_start_min = 0;
+    u8 selection_start_max = 0;
+    u8 selection_end_min = 0;
+    u8 selection_end_max = 0;
+    u8 successive_approximation_low_max = 0;
+    u8 successive_approximation_high_max = 0;
+
+    switch (context.frame.type.type) {
+    case StartOfFrame::FrameType::Type::Baseline_DCT:
+    case StartOfFrame::FrameType::Type::Sequential_DCT:
+        selection_end_min = 63;
+        selection_end_max = 63;
+        break;
+
+    case StartOfFrame::FrameType::Type::Progressive_DCT:
+        selection_start_min = 0;
+        selection_start_max = 63;
+
+        if (spectral_selection_start != 0) {
+            selection_end_min = spectral_selection_start;
+            selection_end_max = 63;
+        }
+
+        successive_approximation_low_max = 13;
+        break;
+
+    case StartOfFrame::FrameType::Type::Lossless:
+        // FIXME: 0 if this is a hierarchical scan
+        selection_start_min = 1;
+        selection_start_max = 7;
+
+        successive_approximation_low_max = 13;
+        successive_approximation_high_max = 15;
+        break;
+    }
+
+    if (spectral_selection_start < selection_start_min || spectral_selection_start > selection_start_max) {
+        dbg() << stream.offset() << String::format(": ERROR! Start of Selection must be in range %i-%i was (%i)", selection_start_min, selection_start_max, spectral_selection_start);
+        return false;
+    }
+
+    if (spectral_selection_end < selection_end_min || spectral_selection_end > selection_end_max) {
+        dbg() << stream.offset() << String::format(": ERROR! End of Selection must be in range %i-%i was (%i)", selection_end_min, selection_end_max, spectral_selection_end);
+        return false;
+    }
+
+    if (successive_approximation_low_bits > successive_approximation_low_max) {
+        dbg() << stream.offset() << String::format(": ERROR! Successive Approximation low bits must be in range 0-%i was (%i)", successive_approximation_low_max, successive_approximation_low_bits);
         return false;
     }
+
+    if (successive_approximation_high_bits > successive_approximation_high_max) {
+        dbg() << stream.offset() << String::format(": ERROR! Successive Approximation high bits must be in range 0-%i was (%i)", successive_approximation_high_max, successive_approximation_high_bits);
+        return false;
+    }
```

Comment:
These validation checks better go in a different function. Just curious...are you planning to add support for progressive JPEG decoding?